### PR TITLE
api: make attestationconfigapi public

### PR DESCRIPTION
**Pull request by elchead**
### Context

The continuum project relies also relies on the attestationconfigapi. So far, the code was duplicated. As agreed upon in chat, the package should be made public to be reused in other projects.

### Proposed change(s)

- move the attestationconfigapi outside of internal
- make the attestationconfigapi cli subpackage internal

### Checklist

- [ ] Run the E2E tests that are relevant to this PR's changes
- [ ] Update [docs](https://github.com/edgelesssys/constellation/tree/main/docs)
- [ ] Add labels (e.g., for changelog category)
- [ ] Is PR title adequate for changelog?
- [ ] Link to Milestone

### CODE REVIEW

**Comment by elchead** [comment]
Testing e2e verify:
https://github.com/edgelesssys/constellation/actions/runs/10788448192

File: api/attestationconfigapi/fetcher.go [hunk 1]
```diff
@@ -18,9 +18,23 @@ import (
...
+
+// Variant is a cloud provider specific attestation variant.
+type Variant interface {
+	String() string
+}
+
 // Fetcher fetches config API resources without authentication.
 type Fetcher interface {
-	FetchLatestVersion(ctx context.Context, attestation variant.Variant) (Entry, error)
+	FetchLatestVersion(ctx context.Context, attestation Variant) (Entry, error)
```
Comment (elchead) [inline api/attestationconfigapi/fetcher.go:37]: continuum for example implements its own variant. I wanted to avoid also exposing this pkg.

File: api/attestationconfigapi/fetcher.go [hunk 2]
```diff
@@ -18,9 +18,23 @@ import (
...
 const cosignPublicKey = constants.CosignPublicKeyReleases
 
+var (
+	// AzureSEVSNP is the Azure SEV-SNP variant.
+	AzureSEVSNP Variant = variant.AzureSEVSNP{}
+	// AWSSEVSNP is the AWS SEV-SNP variant.
+	AWSSEVSNP Variant = variant.AWSSEVSNP{}
+	// GCPSEVSNP is the GCP SEV-SNP variant.
+	GCPSEVSNP Variant = variant.GCPSEVSNP{}
+)
```
Comment (daniel-weisse) [inline api/attestationconfigapi/fetcher.go:28]: Why is this needed?

File: api/attestationconfigapi/fetcher.go [hunk 3]
```diff
@@ -60,7 +74,7 @@ func newFetcherWithClientAndVerifier(client apifetcher.HTTPClient, cosignVerifie
 }
 
 // FetchLatestVersion returns the latest versions of the given type.
-func (f *fetcher) FetchLatestVersion(ctx context.Context, variant variant.Variant) (Entry, error) {
+func (f *fetcher) FetchLatestVersion(ctx context.Context, variant Variant) (Entry, error) {
```
Comment (daniel-weisse) [inline api/attestationconfigapi/fetcher.go:77]:
```suggestion
func (f *fetcher) FetchLatestVersion(ctx context.Context, variant string) (Entry, error) {
```
This might require some more changes to existing code, but why not expect `string` as input?

File: api/attestationconfigapi/fetcher.go [hunk 4]
```diff
@@ -71,31 +85,39 @@ func (f *fetcher) FetchLatestVersion(ctx context.Context, variant variant.Varian
...
 
 // fetchVersionList fetches the version list information from the config API.
-func (f *fetcher) fetchVersionList(ctx context.Context, variant variant.Variant) (List, error) {
-	fetchedList, err := apifetcher.FetchAndVerify(ctx, f.HTTPClient, f.cdnURL, List{Variant: variant}, f.verifier)
+func (f *fetcher) fetchVersionList(ctx context.Context, attestationVariant Variant) (List, error) {
+	parsedVariant, err := variant.FromString(attestationVariant.String())
+	if err != nil {
+		return List{}, fmt.Errorf("parsing variant: %w", err)
+	}
+	fetchedList, err := apifetcher.FetchAndVerify(ctx, f.HTTPClient, f.cdnURL, List{Variant: parsedVariant}, f.verifier)
```
Comment (daniel-weisse) [inline api/attestationconfigapi/fetcher.go:93]:
```suggestion
func (f *fetcher) fetchVersionList(ctx context.Context, variant variant.Variant) (List, error) {
	fetchedList, err := apifetcher.FetchAndVerify(ctx, f.HTTPClient, f.cdnURL, List{Variant: variant}, f.verifier)
```
nit: functions that aren't exported can use our internal type as input. This should also save on some code duplication.

Also applies to `fetchVersion`

**Review by daniel-weisse — COMMENTED** [review]
Can we rename `attestationconfigapi` to just `attestationconfig`?
The package is already in a directory called `api` and the name is needlessly long.

File: api/attestationconfigapi/fetcher.go [hunk 5]
```diff
@@ -18,9 +18,23 @@ import (
...
+	// AWSSEVSNP is the AWS SEV-SNP variant.
+	AWSSEVSNP Variant = variant.AWSSEVSNP{}
+	// GCPSEVSNP is the GCP SEV-SNP variant.
+	GCPSEVSNP Variant = variant.GCPSEVSNP{}
+)
+
+// Variant is a cloud provider specific attestation variant.
+type Variant interface {
+	String() string
+}
```
Comment (msanft) [inline api/attestationconfigapi/fetcher.go:33]: We should drop this and make it an [`fmt.Stringer`](https://pkg.go.dev/fmt#Stringer)

File: api/attestationconfigapi/fetcher.go [hunk 6]
```diff
@@ -71,31 +85,39 @@ func (f *fetcher) FetchLatestVersion(ctx context.Context, variant variant.Varian
 }
 
 // fetchVersionList fetches the version list information from the config API.
-func (f *fetcher) fetchVersionList(ctx context.Context, variant variant.Variant) (List, error) {
-	fetchedList, err := apifetcher.FetchAndVerify(ctx, f.HTTPClient, f.cdnURL, List{Variant: variant}, f.verifier)
+func (f *fetcher) fetchVersionList(ctx context.Context, attestationVariant Variant) (List, error) {
+	parsedVariant, err := variant.FromString(attestationVariant.String())
+	if err != nil {
+		return List{}, fmt.Errorf("parsing variant: %w", err)
+	}
```
Comment (msanft) [inline api/attestationconfigapi/fetcher.go:92]: So when Continuum has its own variants (and they don't exactly match what we have here in Constellation), this will break. I think this package that can only work with a certain set of variants *should* export the package specifying its variants.

**Review by msanft — CHANGES_REQUESTED** [review]

File: api/attestationconfigapi/fetcher.go (hunk 6 above)
Comment (daniel-weisse) [reply api/attestationconfigapi/fetcher.go:92]: Took a look at the code
From what I can tell `variant` is only used to parse strings in this context.
We should just refactor this to work on strings instead of the custom variant type.
We will still receive appropriate error messages if objects do not exist for those strings.

We can provide constants for Constellation's supported variants, e.g. `azure-sev-snp` and `aws-sev-snp`.
I think that was intended by providing [these variables](https://github.com/edgelesssys/constellation/blob/c28e8fbd43c121017f89e28f2d4bf7b8ba796ee9/api/attestationconfigapi/fetcher.go#L21-L28)

File: api/attestationconfigapi/fetcher.go (hunk 2 above)
Comment (elchead) [reply api/attestationconfigapi/fetcher.go:28]: it is not needed, but since it's public a third party should be provided with valid defaults since `attestation/variant` is internal?

File: api/attestationconfigapi/fetcher.go (hunk 5 above)
Comment (elchead) [reply api/attestationconfigapi/fetcher.go:33]: I did this in an earlier commit, but it didn't make the the intention so clear. It doesn't suffice to just pass any stringer object.

File: api/attestationconfigapi/fetcher.go (hunk 3 above)
Comment (elchead) [reply api/attestationconfigapi/fetcher.go:77]: because it requires changing a lot of code

File: api/attestationconfigapi/fetcher.go (hunk 6 above)
Comment (elchead) [reply api/attestationconfigapi/fetcher.go:92]: correct. I only kept this interface to not have to change a lot of places where this is used.

Comment (daniel-weisse) [reply api/attestationconfigapi/fetcher.go:92]: I would rather have changes in a lot of places in this PR, than expose an API that doesn't make sense.
To me, it looks like using the variant interface/type in this package is not needed, so we should remove it.

File: api/attestationconfigapi/fetcher.go (hunk 5 above)
Comment (msanft) [reply api/attestationconfigapi/fetcher.go:33]: Either we make this a string and accept that it will fail late when running into a 404 on the API call (but have a simple interface without many exports), or we export the proper types that can be used here.

**Comment by msanft** [comment]
I'm kind of thinking that this should actually be in its very own repo. If we keep "lending" packages around our projects, it will get messy (circular) at some point.

File: api/attestationconfigapi/fetcher.go (hunk 6 above)
Comment (elchead) [reply api/attestationconfigapi/fetcher.go:92]: I think that the variant interface provides a safer design with the exposed variants, but I can update it to a string.

**Comment by elchead** [comment]
> I'm kind of thinking that this should actually be in its very own repo. If we keep "lending" packages around our projects, it will get messy (circular) at some point.

In the long term yes, but I'm not sure if we should start this for one tiny library now.

File: api/attestationconfigapi/fetcher.go (hunk 6 above)
Comment (daniel-weisse) [reply api/attestationconfigapi/fetcher.go:92]: My issue with the variant interface (which later gets converted to a `variant.Variant` type is, that this only _really_ makes sense in the context of Constellation.
If we expose this package as "public API", then it doesn't make sense to keep this restriction, but hide it behind some named interfaces.
This is especially confusing when other projects also provide their own `Variant` types, that might (or might not) fit into here.

If we just use a string, one can provide whatever they want. If the object does not exist, an error message is returned.
As already mentioned, we can provide the supported values as constants.

**Comment by katexochen** [comment]
What are the stability guarantees the new public API will provide?

**Comment by elchead** [comment]
> What are the stability guarantees the new public API will provide?

The API is backed by an S3 bucket that is versioned (v1). Older SDK versions should always remain compatible with the v1 S3 bucket. Any breaking changes, for newer SDK versions, would be stored in a v2 subpath of the bucket.
wdyt?

**Comment by elchead** [comment]
We decided to not do this and instead copy the code if needed.
